[5.5.x] Add nethealth checker

bernardjkim opened the pull request:
- Backport nethealth #113.
- Add nethealth checker.

Commit:
[5.5.x] Overlay network health testing

Backport nethealth test #113.

bernardjkim committed:
Add nethealth checker

a-palchikov commented on Gopkg.toml:
```diff
@@ -57,7 +60,7 @@
 
 [[constraint]]
-  version = "1.1.7"
+  version = "1.1.8"
```
I think it's safe and recommended to use the latest release (1.1.10) here (and in gravity) as it fixes a number of issues.

a-palchikov commented on cmd/nethealth/main.go:
```diff
@@ -0,0 +1,111 @@
+
+func main() {
+	if err := run(); err != nil {
+		log.Fatal(trace.DebugReport(err))
```
`log.WithError(err).Error("Failed.")` Otherwise the stack trace will be unreadable due to quoting.

a-palchikov commented on cmd/nethealth/main.go:
```diff
@@ -0,0 +1,111 @@
+	var cmd string
+	var err error
+
+	cmd, err = app.Parse(os.Args[1:])
```
nit: technically, you don't need to pre-declare the  variables (unless you were initializing them conditionally):
```go
cmd, err := app.Parse(os.Args[1:])
```

a-palchikov commented on cmd/nethealth/main.go:
```diff
@@ -0,0 +1,111 @@
+
+	cmd, err = app.Parse(os.Args[1:])
+	if err != nil {
+		return trace.Errorf("unable to parse command line.\nUse nethealth --help for help.")
```
`trace.BadParameter`

a-palchikov commented on cmd/nethealth/main.go:
```diff
@@ -0,0 +1,111 @@
+		}
+
+		sigs := make(chan os.Signal, 1)
+		signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
```
I'm wondering why the signal handling here? The default behavior of both SIGINT and SIGTERM is to terminate the process and the handler does not do anything special other than log the signal. Is there maybe a way to trigger graceful shutdown of the server?

bernardjkim replied:
It doesn't look like nethealth currently provides any functionality to gracefully shutdown the server. I'd rather not dig into the nethealth code right now, but I'll take a look into this at a later time.

a-palchikov commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,483 @@
+	}
+
+	if len(echoRequests) != len(echoTimeouts) {
+		return nil, trace.BadParameter("expected equal number of counters for requests and timeouts,"+
```
So I'm still uncertain about this error. Can you clarify how this can be resolved (the comment mentions it might be due to change of labels - does this mean invalid configuration or something else)?

a-palchikov commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,483 @@
+	// defaultNetStatsInterval defines the default interval duration for the netStats.
+	defaultNetStatsInterval = 5 * time.Minute
+
+	// netStatsCapacity sets the number of TTLMaps that can be stored.
```
Not sure I follow this comment. How is this affected by the cluster size being a const?
Also probably needs rephrasing:
```
// netStatsCapacity specifies the maximum number of nethealth samples to store.
```

bernardjkim replied:
Yeah, I'm still trying to figure out how to properly handle this. TTLMap needs to be initialized with a capacity that specifies the max number of elements it can store. This should be equal to the number of members in the cluster. However, the number of members in a cluster is dynamic since members can join/leave. 
Will specify this is just default max number of cluster members for now.

a-palchikov commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,483 @@
+	netStatsTTLSeconds = 60 * 60 * 24 // 1 day
+
+	// packetLossThreshold defines the packet loss percentage used to determine
+	// if overlay network communication with a peer is unhealthy.
```
Needs adding that the loss needs to be consistently observed over the entire interval and not on individual samples

a-palchikov reviewed (approved):
lgtm with a couple of comments

bernardjkim committed:
Bump trace 1.1.8 -> 1.1.10

bernardjkim committed:
Fix nethealth/main.go errors

bernardjkim committed:
Update error message and godoc

bernardjkim committed:
prometheus/client_golang 1.0.0 -> 0.8.0

- Conflict with prometheus/client_golang in teleport

bernardjkim committed:
Revendor gravitational/roundtrip

bernardjkim committed:
Revert "Revendor gravitational/roundtrip"

This reverts commit 823fdbd3fb2b98af312769ca56bc18703fb86cf1.

bernardjkim committed:
Remove roundtrip dependency

r0mant commented on monitoring/nethealth.go:
```diff
@@ -0,0 +1,484 @@
+	return r.TTLMap.Set(peer, data, netStatsTTLSeconds)
+}
+
+// peerData keeps trock of relevant nethealth data for a node.
```
track

r0mant reviewed (approved):
lgtm, I'd be curious to see how this algorithm (calculating timeouts over each interval vs over the entire sample array) performs in the field.

bernardjkim committed:
Update pod field selector to use spec.nodeName

bernardjkim committed:
Increment nethealth request counter

- Increment the request counter even if heart beat fails to send. This
way we don't have the timeout increasing more than the request counter
is increasing.

a-palchikov commented on monitoring/nethealth.go:
```diff
@@ -262,38 +263,59 @@ func (c *nethealthChecker) isHealthy(peer string) (healthy bool, err error) {
-	resp, err := client.Get(ctx, client.Endpoint("metrics"), url.Values{})
+	respCh := make(chan *http.Response, 1)
+	errCh := make(chan error, 1)
+	go func() {
```
I'm seeing this goroutine as an attempt to handle context expiration and HTTP query simultaneously - is this the case?
Why not use a request with a context for that case (as it sort of was before)? Also, using the default HTTP client is not ideal in most of the cases due to its implicit configuration w.r.t timeouts.

bernardjkim replied:
Yeah, I wanted to handle an http request with context here without using the `roundtrip` package. 

Gravity 5.5 is using an older version of `roundtrip`. So if we wanted to use `roundtrip` here we would need to update gravity's version of the package and all of the code that imports the package.

I think gravity 6.x has updated the `roundtrip` package, so we can use the previous implementation in versions >= 6.x

a-palchikov replied:
You don't need roundtrip to make an HTTP request bound to a context:
```go
// create a custom HTTP client
req := http.NewRequestWithContext(ctx)
resp, err := client.Do(req)
```

bernardjkim committed:
Send request with context
